sdk/include/ds/pointer.h

```diff
@@ -0,0 +1,268 @@
+// Copyright Microsoft and CHERIoT Contributors.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file Pointer utilities
+ */
+
+#pragma once
+
+#include <cheri.hh>
+#include <concepts>
+
+namespace ds::pointer
+{
+
+	/**
+	 * Offset a pointer by a number of bytes.  The return type must be
+	 * explicitly specified by the caller.  The type of the displacement offset
+	 * (Offset) is templated so that we can accept both signed and unsigned
+	 * offsets.
+	 */
+	template<typename T, typename Offset = size_t, typename U>
+	static inline __always_inline T *offset(U *base, Offset offset)
+	{
+		CHERI::Capability c{base};
+		c.address() += offset;
+		return c;
+	}
+
+	/**
+	 * Compute the unsigned difference in bytes between two pointers' target
+	 * addresses.  To be standards-compliant, cursor must be part of the same
+	 * allocation as base and at a higher address.
+	 */
+	static inline __always_inline size_t diff(const void *base,
+	                                          const void *cursor)
+	{
+		return static_cast<size_t>(reinterpret_cast<const char *>(cursor) -
+		                           reinterpret_cast<const char *>(base));
+	}
+
+	namespace proxy
+	{
+
+		/**
+		 * Proxies<P,T> if P is a proxy object for T*-s.
+		 */
+		template<typename P, typename T>
+		concept Proxies = std::same_as<T, typename P::Type> &&
+		  requires(P &proxy, P &proxy2, T *ptr)
+		{
+			/* Probe for operator=(T*) */
+			{
+				proxy = ptr
+				} -> std::same_as<P &>;
+
+			/* Probe for operator T*() */
+			{
+				ptr == proxy
+				} -> std::same_as<bool>;
+
+			/* TODO: How to probe for operator-> ? */
+
+			/* Probe for operator==(T*) */
+			{
+				proxy == ptr
+				} -> std::same_as<bool>;
+
+			/* Probe for operator==(P&) */
+			{
+				proxy == proxy2
+				} -> std::same_as<bool>;
+
+			/* Probe for operator<=>(T*) */
+			{
+				proxy <=> ptr
+				} -> std::same_as<std::strong_ordering>;
+
+			/* Probe for operator<=>(P) */
+			{
+				proxy <=> proxy2
+				} -> std::same_as<std::strong_ordering>;
+		};
+
+		/**
+		 * Pointer references are pointer proxies, shockingly enough.
+		 */
+		template<typename T>
+		class Pointer
+		{
+			T *&ref;
+
+			__always_inline void set(T *p)
```

Comment:
This looks redundant.  In the proxies in cheri.hh we use a `set` method like this to simplify subclasses, but this is private (not protected) so looks like it's an unnecessary layer of indirection.